### PR TITLE
Fix position

### PR DESCRIPTION
**Pull request by Mihaiii**
Remove "top: 20px" as it was hardcoded.
Make sure "expandable" variable from "reposition" is the right one.
Make sure "expandedOnElement" is always the same.
Make sure the values are overwritten in case they are changed dynamically.
Add "auto" position in order to make sure "expandable" doesn't get outside browser's window.
Update README.md .

If "expanded" is "left" or acts like it, then its upper-right corner will be near the buttom-right/top-right corner of the "expandedOnElement".
If "expanded" is "right" or acts like it, then its upper-left corner will be near the buttom-left/top-left corner of the "expandedOnElement".
This is the reason I use "elem.scrollWidth" when I set "expandable.style.marginRight".

For https://github.com/StarcounterApps/SignIn/pull/123 .

### CODE REVIEW

File: src/juicy-popover.html [hunk 1]
```diff
@@ -188,7 +188,7 @@
                 expandable.style.right = "0";
                 expandable.style.marginRight = "-" + elem.scrollHeight + "px";
             } else if(position.indexOf("auto") > -1){
-                if(window.innerWidth - elem.getBoundingClientRect().right < expandable.scrollWidth + expandable.style.marginRight) {
+                if(window.innerWidth - elem.getBoundingClientRect().right < expandable.scrollWidth + (parseInt(expandable.style.marginRight) || 0)) {
```
Comment (alshakero) [inline src/juicy-popover.html]: padding radix is twice as fast, `parseInt(expandable.style.marginRight, 10)`.

**Comment by Mihaiii** [comment]
To be honest, I think that the real changes need to be apply somewhere else, blending related. Up to the latest commit, I tested it on SignIn app using Website app and SampleWebsiteTheme. The latest commit was made after I tried to test it in Calendar app and, as you can see, the 2 cases are uite different.

File: src/juicy-popover.html [hunk 2]
```diff
@@ -163,23 +162,59 @@
 
         JuicyPopoverElementPrototype.reposition = function () {
             var expandable = this.shadowRoot.querySelector("#expandable");
+            var expandableSlot = null;
```
Comment (tomalec) [inline src/juicy-popover.html]: Regarding naming, I would say, the one above is a `-Slot` this one is just `expandable` element, isn't it?

File: src/juicy-popover.html [hunk 3]
```diff
@@ -163,23 +162,59 @@
...
+                    styleTarget.style.right = "";
+                    styleTarget.style.marginRight = "";
+                }
             }
         };
 
         JuicyPopoverElementPrototype.expand = function (elem) {
+            if (!elem.assignedSlot && this.attachShadow) {
+                elem = elem.parentElement;
+            }
```
Comment (tomalec) [inline src/juicy-popover.html]: Why this is needed?

**Comment by tomalec** [comment]
> To be honest, I think that the real changes need to be apply somewhere else, blending related.

I don't quite get it? What exactly should go somewhere else and why?

**Comment by Mihaiii** [comment]
> What exactly should go somewhere else and why?
That's what I don't know, but I'm not really comfortable with this PR because it feels like a hack. I don't understand why in some cases the `clientWidth`/`clientHeight` is in the light DOM and in others in the shadow DOM.

Just as a side note, as I just realize it: with this changes, using "top" will display the popover on the top of the handler, rather than in the sideways as it was before. I will let "top" the way is was before and make a new option, if you don't mind.

File: src/juicy-popover.html (hunk 3 above)
Comment (Mihaiii) [reply src/juicy-popover.html]: You're right, I forgot to erase to it.
FIxed in 913fd7233b3029a9113e958c0d50458e814d2ac9 .

File: src/juicy-popover.html (hunk 2 above)
Comment (Mihaiii) [reply src/juicy-popover.html]: Thanks!
Renamed in 57ad7e2b0db98bc74825e2ab49530c17639492bf .

**Comment by warpech** [comment]
I will continue the work on this once https://github.com/Juicy/juicy-popover/pull/15 is merged.

**Comment by warpech** [comment]
I have reworked this PR. Please check README.md and the commit message for description. Now it does not need taking Shadow DOM into consideration while doing `reposition`.

The default version should be good enough for Sign In, but actually the fallback position behavior is now completely customisable.

@PatrykSzwer and @Mihaiii, can you pls review?

**Comment by Mihaiii** [comment]
It looks great in SignIn app.
I tried it in Calendar app also and the position of the popover is a strange.
calendar

**Comment by Mihaiii** [comment]
This is the demo page with latest changes:
popover

**Comment by warpech** [comment]
> I tried it in Calendar app also and the position of the popover is strange.

It is because Calendar view has `<juicy-popover slot="calendar/filters-popover" position="bottom">` . The position `bottom` is really interpreted as "bottom right". Should I change the default behavior to be "bottom left"?

 Or do you think this should be corrected in the Calendar app: `<juicy-popover slot="calendar/filters-popover" position="bottom left">` or even better `<juicy-popover slot="calendar/filters-popover">`.

**Comment by warpech** [comment]
> Should I change the default behavior to be "bottom left"?

I will change this in `juicy-popover`, otherwise it is an unceccessary breaking API change.

**Comment by warpech** [comment]
Fix for Calendar pushed. Regarding Sign In, I cannot reproduce it locally. Hopefully, this is a caching issue on your side?

**Comment by Mihaiii** [comment]
Yes, regarding my previous comment, it was my mistake (I was using another branch).
I checked in Calendar also and look ok with me.

From my part, this can be merged.

**Comment by warpech** [comment]
Thank you, I will merge, release a version and ping you in https://github.com/StarcounterApps/SignIn/pull/123

**Pull request merged**